### ISSUE TITLE
Error trying to run openifs 48r1 standalone.

### ISSUE THREAD

**Issue by JanStreffing** [esm-tools/esm_tools#1204]
This used to work, but does not seem to work anymore on the release branch.

**Describe the bug**
```
Copying standard yamls from: /home/a/a270092/esm_tools/configs/
Copying standard namelists from: /home/a/a270092/esm_tools/namelists/
First year, checking if we need to compile...

================================================================================
STARTING SIMULATION JOB!
Experiment ID = 48r1_debug1
Setup = oifs
Experiment is installed in:
       /work/ab0246/a270092/runtime/oifs-48r1//48r1_debug1
================================================================================
Traceback (most recent call last):
  File "/home/a/a270092/.local/bin/esm_runscripts", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_runscripts')())
  File "/home/a/a270092/esm_tools/src/esm_runscripts/cli.py", line 289, in main
    setup()
  File "/home/a/a270092/esm_tools/src/esm_runscripts/sim_objects.py", line 133, in __call__
    self.prepcompute()
  File "/home/a/a270092/esm_tools/src/esm_runscripts/sim_objects.py", line 224, in prepcompute
    self.config = prepcompute.run_job(self.config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepcompute.py", line 26, in run_job
    config = evaluate(config, "prepcompute", "prepcompute_recipe")
  File "/home/a/a270092/esm_tools/src/esm_runscripts/helpers.py", line 71, in evaluate
    config = esm_plugin_manager.work_through_recipe(
  File "/home/a/a270092/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 159, in work_through_recipe
    config = getattr(submodule, workitem)(config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepcompute.py", line 271, in _write_finalized_config
    config.yaml_dump(config_file_path)
  File "/home/a/a270092/esm_tools/src/esm_parser/dict_to_yaml.py", line 117, in yaml_dump
    my_yaml.dump(config_final, stream)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/main.py", line 567, in dump
    return self.dump_all([data], stream, transform=transform)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/main.py", line 577, in dump_all
    self._context_manager.dump(data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/main.py", line 913, in dump
    self._yaml.representer.represent(data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 81, in represent
    node = self.represent_data(data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 103, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 316, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 842, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 103, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 316, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 842, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 113, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/home/a/a270092/.local/lib/python3.9/site-packages/ruamel/yaml/representer.py", line 345, in represent_undefined
    raise RepresenterError(f'cannot represent an object: {data!s}')
ruamel.yaml.representer.RepresenterError: cannot represent an object: (1+0j)
Traceback (most recent call last):
  File "/home/a/a270092/.local/bin/esm_runscripts", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_runscripts')())
  File "/home/a/a270092/esm_tools/src/esm_runscripts/cli.py", line 289, in main
    setup()
  File "/home/a/a270092/esm_tools/src/esm_runscripts/sim_objects.py", line 119, in __call__
    self.config = prepexp.run_job(self.config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepexp.py", line 27, in run_job
    evaluate(config, "prepexp", "prepexp_recipe")
  File "/home/a/a270092/esm_tools/src/esm_runscripts/helpers.py", line 71, in evaluate
    config = esm_plugin_manager.work_through_recipe(
  File "/home/a/a270092/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 159, in work_through_recipe
    config = getattr(submodule, workitem)(config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepexp.py", line 260, in call_esm_runscripts_from_prepexp
    _call_esm_runscripts_internally(config, new_command, scriptsdir)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepexp.py", line 193, in _call_esm_runscripts_internally
    subprocess.check_call(command.split(), cwd=exedir)
  File "/sw/spack-levante/mambaforge-4.11.0-0-Linux-x86_64-sobz6z/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['esm_runscripts', 'oifs-48r1-levante-TCO95L91.yaml', '-e', '48r1_debug1', '--open-run', '--no-motd', '--last-jobtype', 'prepcompute', '-t', 'prepcompute']' returned non-zero exit status 1.
```

**To Reproduce**
```
 ╭─◀ 🖥  LEVANTE a270092 ▶ ~/esm_tools/runscripts/oifs ▶ 🔀 release ▶
 ╰❯ $ esm_runscripts oifs-48r1-levante-TCO95L91.yaml -e 48r1_debug1 -U --open-run
```

**Comment by pgierz** [comment]
A `ruamel` error. That's a first. Have you tried this since we added the provenance feature, or is this the first time that this error is happening since it went in?

**Comment by pgierz** [comment]
Error is reproducible. What exactly are you trying to dump?

```
ruamel.yaml.representer.RepresenterError: cannot represent an object: (1+0j)
```

Maybe a scientific number? Does oifs have any weird types in its config?

**Comment by pgierz** [comment]
Interesting. Why do you have complex numbers?
```
data = config_final
import dpath.util
all_values = list(dpath.util.values(data, "**"))
{<class 'tuple'>, <class 'dict'>, <class 'esm_runscripts.batch_system.batch_system'>, <class 'str'>, <class 'bool'>, <class 'int'>, <class 'esm_calendar.esm_calendar.Calendar'>, <class 'esm_calendar.esm_calendar.Date'>, <class 'complex'>, <class 'NoneType'>, <class 'list'>}
```

With some small Python wizardry:
```python
>>> import dpath.util
>>> def find_paths_and_values_by_type(data: dict, target_type: type):
...     # Get all paths and values in the nested dictionary
...     all_paths_and_values = list(dpath.util.search(data, '**', yielded=True))
...
...     # Filter by the target type
...     filtered_paths_and_values = [(path, value) for path, value in all_paths_and_values if isinstance(value, target_type)]
...
...     return filtered_paths_and_values
...
>>> data = config_final
>>> target_type = complex
>>> paths_and_values = find_paths_and_values_by_type(data, complex)
>>> for path, value in paths_and_values:
...     print(f"Path: {path}, Value: {value}")
...
Path: oifs/nrest_ind, Value: (1+0j)
```

I'm guessing it's this one:
https://github.com/esm-tools/esm_tools/blob/d8bc2ff6369966555f2dc4ef4eac10cf614db1b4/configs/components/oifs/oifs.yaml#L1107

**Comment by JanStreffing** [comment]
idk, that line has been there for 3 years:
https://github.com/esm-tools/esm_tools/commit/9f1da6030123c70389d83005a8c74f5b6740e527
Something change with the handling of that variable internally?

**Comment by pgierz** [comment]
Very probably, yes. We use a new dumper now to accommodate the provenance comments, and somewhere it is converting strings to complex numbers if they fit a certain pattern. No idea where that happens, yet. I'll keep looking into it.

**Comment by pgierz** [comment]
TLDR: changed the parser to not allow complex numbers. I see no sensible scenario where our tools need complex numbers anyway. Try branch [`fix/parser_complex`](https://github.com/esm-tools/esm_tools/pull/1206). If it works extensively enough for you I'll put it in.

- - - -

You managed to hit an edge case that has been there for (more than) 3 years, has a bug in it, dates back from the very unpleasant hacking together of the original parser...probably originally from when we merged everything into one package from all the separate ones. Or even earlier.

https://github.com/esm-tools/esm_tools/blob/d8bc2ff6369966555f2dc4ef4eac10cf614db1b4/src/esm_parser/esm_parser.py#L2572-L2577

First off, this function is just wrong (or, if it does what @dbarbi had in mind, then it's badly named):
https://github.com/esm-tools/esm_tools/blob/d8bc2ff6369966555f2dc4ef4eac10cf614db1b4/src/esm_parser/esm_parser.py#L2580-L2584

The right way is:

```python
def contains_underscore(value):
    if insinstance(value, str):
        return "_" in value
    # Not a string, so it can't have an underscore
    return False
```

(Fixing it causes **everything** to break, so I'm not touching that for right now even if someone paid me an exorbitant amount of money)

So, we have:
```python
bad_value = "(1)"
try: 
    complex(value) 
    return contains_underscore(value) # <-- gives back True
except: 
    return False
```

You get back that the number can be complex, it is converted to a complex number, and subsequently dumping results in an error.

**Comment by JanStreffing** [comment]
ty!
